### ISSUE TITLE
Pastel: disable colors on non-tty output

### ISSUE THREAD

**Issue by Schniz** [facebookexperimental/reason-native#100]
Hey! 👋 
first of all, I'd like to say that this repo is amazing. Thank you all for your hard work. I've tried Pastel and Rely and they both work very good 🏆 

With Pastel is able to disable colors using a different `mode`, I believe we can make a function that mutates the mode or uses `useMode` and knows about the tty. So the following would work like:

* `./pastel-program` would be colorized
* `./pastel-program | cat` would not be colorized
* `echo $(./pastel-program)` would not be colorized

that would also help for things like Neovim or non-tty hosts of Rely, where this is the current behavior:

![image](https://user-images.githubusercontent.com/2054772/52621255-7cd50300-2eaf-11e9-8976-7eb8e51dc7dc.png)

**Comment by kyldvs** [comment]
Thanks, and nice idea!

It looks like we'd use https://caml.inria.fr/pub/docs/manual-ocaml/libref/Unix.html#VALisatty

A few questions we'd want to decide on:

1. Should this functionality go in Pastel or in the libraries using Pastel (like Rely)?
2. If it's in Pastel should it be on or off by default? (With an easy way to switch to tty detection).
3. Are there any concerns adding a dependency on Unix?

**Comment by Schniz** [comment]
> 1. Should this functionality go in Pastel or in the libraries using Pastel (like Rely)?

Maybe we can make a module similar to [`supports-color`](https://github.com/chalk/supports-color/blob/master/index.js), providing all kinds of ways to determine if you support color (including tty, `FORCE_COLOR` env var, etc). Then it's easy as:

```reason
Pastel.setMode(
  SupportsColor.supports()
  ? Pastel.Terminal
  : Pastel.Disabled
)
```

**Comment by Schniz** [comment]
> If it's in Pastel should it be on or off by default? (With an easy way to switch to the non-default tty detection mode).

I agree that it would be weird to be on by default. Pastel generates strings, it doesn't feel like something that checks `istty` or environment variables. `Pastel.inferMode()` can also be an option

**Comment by bandersongit** [comment]
First off, thanks for filing, and I'm glad to hear that Rely and Pastel have been working well for you.

> > 1. Should this functionality go in Pastel or in the libraries using Pastel (like Rely)?
> 
> Maybe we can make a module similar to [`supports-color`](https://github.com/chalk/supports-color/blob/master/index.js), providing all kinds of ways to determine if you support color (including tty, `FORCE_COLOR` env var, etc). Then it's easy as:
> 
> ```reason
> Pastel.setMode(
>   SupportsColor.supports()
>   ? Pastel.Terminal
>   : Pastel.Disabled
> )
> ```

I had the exact same library in mind :p, I think we want to be exposing at least supportsColor(file_descr) because Pastel doesn't know what you will be doing with your strings. 

> > If it's in Pastel should it be on or off by default? (With an easy way to switch to the non-default tty detection mode).
> 
> I agree that it would be weird to be on by default. Pastel generates strings, it doesn't feel like something that checks `istty` or environment variables. `Pastel.inferMode()` can also be an option

I think that being on by default makes things much simpler for consumers (and I understand this sort of contradicts what I said above, but I think that assuming stdout will in most cases be helpful). I can see issues like the one you filed that references this cropping up once we add this feature because people won't know to use inferMode() and if it works on their machine, they won't think to look for it.

Either way, I think this will constitute a breaking change and will require a major version bump. Right now Pastel exposes a top level implementation (which I think anyone writing libraries for external consumption probably wants to use) as well as a functor for making sandboxed implementations (kind of niche, we mostly use it for testing, but I can see other applications). In the functor world I think we can take a similar approach to what we have done in Rely and have our argument to the functor be something like the below (and have inferMode be on by default).

```reason
module type PastelConfiguration = {
  let config: t;
  let initialize: unit => t;
  let inferMode: (bool, t) => t;
}

module Make: (Config: { let config: PastelConfiguration.t }) => Pastel.Sig
```

With this approach I think that the inferMode flag to the functor would cause Pastel.defaultMode to be inferred instead of just being TerminalMode.

**Comment by kyldvs** [comment]
> I think that being on by default would constitute a breaking change and require a major version bump, but I think that the user experience improvement is worth it.

We can probably add a non-breaking change that allows turning on the color inference but does not change default behavior. Then we can decide in the future if we want to make that the default with a breaking change.

**Comment by bandersongit** [comment]
> > I think that being on by default would constitute a breaking change and require a major version bump, but I think that the user experience improvement is worth it.
> 
> We can probably add a non-breaking change that allows turning on the color inference but does not change default behavior. Then we can decide in the future if we want to make that the default with a breaking change.

Thinking about it more, I think this could also reasonably just be characterized as a bugfix because ansi escape codes are just kind of meaningless/clearly undesirable in the wrong context. If we can buy that argument then I think we can make it the default (which I still think is better) without a major version.

Also I'm going to implement TTY checking (and probably the environment variable as well) without changing the functor in the next day or so.

**Comment by kyldvs** [comment]
> If we can buy that argument then I think we can make it the default (which I still think is better) without a major version.

I don't think the barrier to a major version bump should be that high. If something is reasonably considered a breaking change (which I think this is) we can push it out and do a major bump. It doesn't need to have anything else included or a lot of fanfare around it IMO.

**Comment by bandersongit** [comment]
Just released Pastel 0.2.0 which includes this feature. I will release Rely 1.2.0 later today which updates the default Rely reporter to provide pretty output in this case among other things as well.

It is worth calling out that you may need to do something like the below to get the behavior you want because ```let someText = <Pastel color=Red>"Some Text"</Pastel>```
can get optimized into a constant by the compiler in whatever environment you compile in.
In the case of Rely changing to ```let someText = () => <Pastel color=Red>"Some Text"</Pastel>``` was sufficient, but I am not entirely sure why the RHS wasn't optimized.
